core/domain/expression_parser.py

```diff
@@ -0,0 +1,533 @@
+# Copyright 2014 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A parser that generates a parse tree for math expressions.
+
+It uses the following grammar in Backus-Naur form:
+
+<expr> ::= <mul_expr> (('+' | '-') <mul_expr>)*
+<mul_expr> ::= <pow_expr> (('*' | '/') <pow_expr>)*
+<pow_expr> ::= '-' <pow_expr> | '+' <pow_expr> | <unit> ('^' <pow_expr>)?
+
+<unit> ::= <identifier> | <number> | '(' <expr> ')' | <function> '(' <expr> ')'
+<number> ::= r'[0-9]+.[0-9]+|[0-9]+'
+<identifier> ::= r'[a-zA-Z]' | 'alpha' | 'beta' | 'gamma' | 'theta' | 'epsilon'
+| 'pi' | 'omega'
+<function> ::= 'sqrt' | 'abs' | 'cos' | 'sin' | 'tan' | 'cot' | 'sec' | 'cosec'
+"""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import re
+
+import python_utils
+
+GREEK_LETTERS = ['alpha', 'beta', 'gamma', 'theta', 'epsilon', 'pi', 'omega']
+MATH_FUNCTIONS = ['sqrt', 'abs', 'cos', 'sin', 'tan', 'cot', 'sec', 'cosec']
+VALID_OPERATORS = ['[', '{', '(', ')', '}', ']', '+', '-', '/', '*', '^']
+
+
+def contains_balanced_brackets(expression):
```

Comment:
@iamprayush I think I've mentioned this before, but please note that every public function needs to have its own unit tests.

Reply:
(Please check that this is the case for all functions in this PR.)

Reply:
@seanlip I'll add the tests for `contains_balanced_brackets` and `is_algebraic` since they are global functions. But just to confirm, I don't need to add tests for the parser functions, right? Since all of them, are being checked by the test cases that I've already added. 
Also, I'm not sure how exactly they would be tested since they are interdependent recursive functions that implement a particular production rule of the grammar, so having separate independent unit tests for them doesn't really make sense since none of them can ever be used independently.

Reply:
Tests for global functions added.

Reply:
Tests added.